substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -1197,8 +1202,20 @@ where
 									IncomingBlock {
 										hash: block_data.block.hash,
 										header: block_data.block.header,
-										body: block_data.block.body,
-										indexed_body: block_data.block.indexed_body,
+										// If block pruning is enabled, the user does not care about
+										// historical data. Don't store block bodies, only store
+										// headers since they're necessary for correct node
+										// operation.
+										body: if self.block_pruning_enabled {
```

Comment:
The filtering here is happening too late, the block bodies have already been fetched. In https://github.com/paritytech/polkadot-sdk/issues/5119 you can see that we define the attributes to fetch [here](https://github.com/paritytech/polkadot-sdk/blob/c158d5d16b58f30a214fbfba278b8bad4daa70ef/substrate/client/network/sync/src/strategy/chain_sync.rs#L1541-L1541). I think we can just check whether gap sync is active and then not even fetch the bodies. 

Since we know that all the bodies take hundreds of gigabytes of disk space, discarding them here means that we still download all the gigabytes of bodies.